src/picongpu/include/particles/Particles.kernel

```diff
@@ -255,93 +255,186 @@ struct KernelManipulateAllParticles
-
-       const DataSpace<simDim> blockCell = block * SuperCellSize::toRT();
-
-
-
-       FramePtr frame;
-
-       PMACC_SMEM( mustShift, int );
-       lcellId_t particlesInSuperCell;
-
-
-       if (linearThreadIdx == 0)
-       {
-           mustShift = 0;
-       }
-       frame = pb.getLastFrame(block);
-       particlesInSuperCell = pb.getSuperCell(block).getSizeLastFrame();
-
-       auto cachedB = CachedBox::create < 0, typename BBox::ValueType > (BlockDescription_());
-       auto cachedE = CachedBox::create < 1, typename EBox::ValueType > (BlockDescription_());
-
-       __syncthreads();
-       if (!frame.isValid())
-           return; //end kernel if we have no frames
-
-
-       auto fieldBBlock = fieldB.shift(blockCell);
-
-       nvidia::functors::Assign assign;
-       ThreadCollective<
-           BlockDescription_,
-           PMacc::math::CT::volume< typename BlockDescription_::SuperCellSize >::type::value
-       > collective( linearThreadIdx );
-       collective(
-                 assign,
-                 cachedB,
-                 fieldBBlock
-                 );
-
-       auto fieldEBlock = fieldE.shift(blockCell);
-       collective(
-                 assign,
-                 cachedE,
-                 fieldEBlock
-                 );
-       __syncthreads();
-
-       /*move over frames and call frame solver*/
-       while (frame.isValid())
-       {
-           if (linearThreadIdx < particlesInSuperCell)
-           {
-               frameSolver(*frame, linearThreadIdx, cachedB, cachedE, mustShift);
-           }
-           frame = pb.getPreviousFrame(frame);
-           particlesInSuperCell = PMacc::math::CT::volume<SuperCellSize>::type::value;
-
-       }
-       __syncthreads();
-       /*set in SuperCell the mustShift flag which is a optimization for shift particles and fillGaps*/
-       if (linearThreadIdx == 0 && mustShift == 1)
-       {
-           pb.getSuperCell(mapper.getSuperCellIndex(DataSpace<simDim > (blockIdx))).setMustShift(true);
-       }
+       T_ParBox pb,
+       T_EBox fieldE,
+       T_BBox fieldB,
+       T_ParticleFunctor particleFunctor,
+       T_Mapping mapper
+    ) const
+    {
+        using namespace mappings::threads;
+
+        constexpr uint32_t frameSize = PMacc::math::CT::volume< SuperCellSize >::type::value;
+        constexpr uint32_t numWorkers = T_numWorkers;
+
+        uint32_t const workerIdx = threadIdx.x;
+
+        typedef typename T_ParBox::FramePtr FramePtr;
+
+        DataSpace< simDim > const block(
+            mapper.getSuperCellIndex( DataSpace< simDim >( blockIdx ) )
+        );
+
+        // relative offset (in cells) to the supercell (including the guard)
+        DataSpace< simDim > const superCellOffset = block * SuperCellSize::toRT();
+
+         using ParticleDomCfg = IdxConfig<
+            frameSize,
+            numWorkers
+        >;
+
+        PMACC_SMEM(
+            mustShift,
+            int
+        );
+
+        /* unique for each worker */
```

Comment:
we will add an other comment on how it's processed below - but this comment here is confusing for readers since:
- the pointer is stored in a local var is obvious from the definition
- if that means the frame will be uniquely processed or not is not meant by this comment